### ISSUE TITLE
philosophical discussion : getting the wrong answer quickly

### ISSUE THREAD

**Issue by JohnDenker** [atsepkov/RapydScript#93]
By way of background:   I have a 600-line python program that I recently ported from plain python to rapydscript.  This convinced me that rapydscript is a Good Thing. 

Also:  Some things are a matter of opinion, and a matter of taste.  De gustibus non disputandum.  However, without meaning to dispute matters of opinion, each person should be allowed to express an opinion, within limits.

Executive summary:  I don't mind if it's incompatible, to the extent necessary.  I just don't want it to be ***insidiously*** incompatible.

------------------------------

I understand that there is a tradeoff between runtime efficiency and python compatibility.  As I see it, there are at least three possibilities:

1. Do the compatible thing, no matter what the cost.
2. Do the efficient thing, no matter what the cost.
3. Throw an error, in cases where compatibility would be unduly inefficient.

Speaking for myself, I would prefer option 3.  To say the same thing another way:  ***I do not need a tool for giving me the wrong answer quickly.***

There is a rather large segment out there consisting of people who know python but don't know javascript and don't want to learn javascript.  This may not be the segment you choose to serve, but it does exist.  They need a compiler that will protect them from the creepiness of javascript, insofar as possible.

As a related matter, there are people out there who have a tremendous investment in existing python code.  They want to protect this investment.  They don't want to rewrite everything from scratch.

For me, here's where the rubber meets the road.  When I ported my 600-line program from python to rapydscript, one of the very, very few things that I had to change was array equality, i.e. testing (x,y)==(X,Y).

So let's talk about that.  I reckon it would be better to provide two functions, or rather two macros that look like functions:

  a. weq( (x,y),  (X,Y) )        # weird pseudo-equality or weak pseudo-equality
  b. eq( (x,y),  (X,Y) )          # real equality

These compile to javascript as follows:

  a. [x,y] == [X,Y]               # super-efficient, but gives the wrong answer, completely useless AFAICT
  b. eq( [x,y], [X,Y]  )         # slow, but gives the right answer
  .  .  .   In common cases, the previous line could be optimized to:
  b'.  (x==X and y==Y)        # reasonably efficient, and gives the right answer

If there is some deep reason why eq() must be implemented as a function call, rather than a macro that generates inline code, I can accept that ... but even then, I should be given the choice of whether to call eq() or weq().  Sometimes I want convenience, and sometimes I need efficiency.  On Tuesday I might want one thing, and on Wednesday I might need the other.  I don't know in advance which I am going to need ... and if I don't know, the compiler cannot possibly know.

What I don't need and don't want is insidious incompatibility.  I don't want something that looks like python but insidiously returns a non-pythonic answer.

To repeat:  If you want my opinion:

* Efficiency is good.
* Compatibility is good.
* I understand that sometimes there's a conflict.
* Insidious incompatibility is bad.  If it has to be incompatible, it should "loud and proud".  That is, it should be openly, conspicuously, unmistakably incompatible.  If possible, it should throw an error at compile time.  Failing that, it should throw an error an run time.  It should never, ever return the wrong answer without warning.  ***I do not need a tool for giving me the wrong answer quickly.***  I already have plenty of ways of getting the wrong answer.

Remember Knuth's dictum
>   **Premature optimization is the root of all evil.**

In my opinion, mapping "python ==" to "javascript ==" (for non-scalars) is an unhelpful optimization.  It's insidious.
![trap](https://cloud.githubusercontent.com/assets/14043887/9862716/8cde798a-5aef-11e5-9e28-e9a92e6a3392.jpeg)

**Comment by hartsantler** [comment]
"There is a rather large segment out there consisting of people who know python but don't know javascript and don't want to learn javascript."

learning javascript can not be avoided. in any real-world project, you are going to be interfacing with large amounts of external javascript libraries and code, so learning how js works simply can not be avoided.

**Comment by atsepkov** [comment]
@JohnDenker I understand where you're coming from, but be mindful of other use cases, not just your own. The 13,000% slowdown (yes, I got the zeroes right) I mentioned in the other thread is nothing to sneeze at.

1. Consider the [Paint](https://github.com/atsepkov/RapydScript/tree/master/examples/paint) application, for example. It uses a flood-fill algorithm for its paint-bucket tool. The algorithm performs just as fast as the JavaScript version, and appears to be instantaneous to the user. This same logic, originally written by me for Pyjamas to showcase my HTML5Canvas implementation, would freeze the browser for several seconds. The main reason for the slowdown was support for operator overloading that Pyjamas had. When you're dealing with pixels, speed matters. Similarly, if you are feeling outraged by having to use `eq()`, why should `weq()` be forced on another unaware user? Why should WebGL-based [glowscript](http://www.glowscript.org/#/user/GlowScriptDemos/folder/Examples/) or [Ray-Tracing](http://salvatore.pythonanywhere.com/RapydRay) suffer? And what's more commonly compared, arrays or primitives? If your 600 lines of Python would all of a sudden halt to a crawl with the first execution, would you continue to use RapydScript or would you assume that it's broken (or is another educational tool with no use in the real world)? Be honest. Having the app freeze with no answer at all is no better than getting the wrong answer - and wrong answer is easier to debug.

2. RapydScript is not the only language where `==` operator only performs identity comparison. Some other popular languages that do the same include Java, C, C++, Objective-C and Go. Being aware of concept of deep vs shallow comparison is a fact of life in programming. Afaik, the automatic deep comparison in Python is a result of each object implementing their own version of `__eq__`. You may be disappointed to find that operator overloading in RapydScript doesn't exist either because it suffers from the same problem.

3. As hartsantler mentioned, if you plan to program for the web, you can only escape JavaScript for so long. RapydScript is meant to improve readability and make it easier to write maintainable code rather than to encourage apathy towards the underlying platform. As with all languages, you can only get so far without learning advanced features. There is a whole laundry list of differences in JavaScript world, just the concept of asynchronous programming is a puzzle in itself.

4. Your optimization is a bit naive. What if `x` and `X` are also arrays? Then we're back to where we started. Moreover, the problem is not being able to compare arrays fast, the problem is being able to compare primitives fast (especially when compiler doesn't have enough information to decide when `eq` can be replaced with `==`), because you'll be seeing a lot of them in graphic environments. I'm not saying your approach won't work, you may be on to something here, and if you want to tackle this, you're welcome to. Optimizations in general are time-consuming and there is a whole task dedicated to optimizer enhancements.

5. Rather than complain on your first day about a feature, why don't you give the language more time? Many users have come and went, many users asked about features they no longer cared about a week later. I'm not going to break the language to accommodate a user who may not even be around tomorrow. Besides, it's not like the logic you're asking for isn't in the language, it's here, it's just called `eq()` rather than `==`.

With that said, I do have a trick up my sleeve in the pipeline to improve equality testing, but I'll wait with it until I'm done merging bug fixes from -ng fork. There are a lot of good fixes Kovid added, and I'd like other members in the main branch to benefit from them.

**Comment by kovidgoyal** [comment]
Given the position RS occupies on the speed vs compatibility tradeoff (a position I agree with), there are simply too many incompatibilities between RS and python to ever make direct use of non-trivial python code feasible -- and I say that as some that has written half a miliion lines of python. I would like nothing better than to just copy paste python code, or even directly import python modules, but I think we have to wait for webassembly to gain a few things like garbage collection before doing something like this becomes even remotely possible.

**Comment by JohnDenker** [comment]
Let's talk about this.  I'm not saying this is a bug, or even a creeping feature request.  It is, however, something worth discussing.

  1. As a first example (not the only example):  When I am using python to calculate the cooking time for some hard-boiled eggs, I don't care whether the calculation takes 1 nanosecond or 130 nanoseconds.  I am much more interested in getting the right answer.  If I have to spend a few minutes debugging it, because the "==" operator did the wrong thing, that wastes a whole lot more than 130 nanoseconds.
  2. More generally, I find that real-world code consists of vast, vast sections that are not time-critical, with at most a few small sections that are.  Optimizing the non-critical sections is not worth the trouble.  I am not the only person to think so.  Some guy named Knuth said ***Premature optimization is the root of all evil***.
  3.  Since you asked, yes, I am being honest about this.  Actually.  Really.  Truly.
  4.  The efficiency of glowscript was mentioned.  Glowscript in particular and visual python in general do a tremendous amount of operator overloading.  This allows addition of vectors using the "+" operator, multiplication of vectors by scalars using the "*" operator, and various other things.  So this example does not prove that high-level language operations are inconsistent with acceptable efficiency;  in fact it pretty much proves the opposite.
  5.  The C++ language was mentioned.  In C++ the variable-types are known at compile time.  This makes it relatively easy to optimize things like the "==" operator:  If the variables are ints, generate the int comparison inline, with no overhead whatsoever.  If the variables are nested lists, invoke the appropriate function.  If the variable types are incompatible, generate an error at compile time (not run time).  Again, this example does not prove that high-level language operations are inconsistent with acceptable efficiency;  in fact it pretty much proves the opposite.
  6.  As always, when I argue against one extreme, it does ***not*** mean that I am arguing in favor of the opposite extreme.  Usually all the extremes are wrong.  For example, in the present discussion:
    * I am not advocating efficiency without regard to compatibility.
    * I am also not advocating compatibility without regard to efficiency.
    * I am suggesting that there is a middle way.  Please let's discuss the middle way.  Please let's not harangue one another about the evils of one extreme or the other.
  7.  The question was asked, what if x and y are also arrays?  Two answers:
   * Minor point:  If we take the question literally:  There are ways of recursively exploring a tree without doing a lot of function calls.  Functional programming has many advantages, especially as a conceptual model ... but in *compiled assembly code* on a virtual machine where function-calls are expensive, it is not the only way of doing things.
   * Larger point:  Responding more to the spirit of the question:  Suppose X and Y are objects for which comparison is necessarily expensive. This can be detected at runtime!   In the worst case, check the type of X, check the type of Y, and evaluate X==Y if that makes sense;  otherwise throw an exception.  I would hope that type-checking slows things down by *at most* a factor of 2 or 3, not a factor of 130.
  8. As an example of the best sort of optimization, consider the following code, which fills in the diagonal of a matrix.  In this case -- not in all cases, but in this case -- it is known *at compile time* that ii and jj are integers, so the compiler can emit "javascript ==" with no runtime checking, no slowdown whatsoever, and no risk of getting the wrong answer.
```
        for ii in range(size):
          for jj in range(size):
            if ii==jj:  M[ii][jj] = 1
```
  9. Here's a trickier case:  Suppose X and Y are formal parameters within a function, passed in from who-knows-where, such that their type cannot be known at runtime.  The programmer should be allowed to ASSERT that X and Y are ints.  This would allow the compiler to generate highly optimized code, without runtime checks in the critical section, and without the risk of getting the wrong answer.
```
       assert X is an integer
       assert Y is an integer
       for ii in range(huge):
          for jj in range(huge):
              if X == Y:   #.... some time-critical code ....#
```
  10.  When I say such-and-such is my opinion, it means it is my opinion.  When I say that this-or-that market segment exists, it means it exists;  it does not mean it is the only segment.  It is really not necessary to tell me that other segments exist.  Indeed that shoe is on the other foot:  We agree that there is a *segment* that necessarily needs to be proficient in both python and javascript, but I insist that it is ***not*** the only segment.  In particular, python is widely used in education, as students' first language.  These students barely have enough time and energy to learn one language; they are certainly not going to learn two, not on the relevant timescale.  Also, these students greatly benefit from something that runs in the browser, so they don't need to download and install some heavy-duty development environment.  As I said before, this may not be the segment you choose to serve, but I insist it does exist.

These students ***do not need a tool for getting the wrong answer quickly.***  They do not care whether the == operator takes 1 nanosecond or 130 nanosecond.  For them, that's not the rate-limiting step.

Bottom line:  Compiling "python ==" into "javascript ==" may be convenient for the compiler, but that doesn't make it right.  I find the efficiency arguments that have been made so far to be less than compelling, insofar as they bang from one Manichean extreme to the other, without considering the middle ground.  There may be a good reason out there somewhere, but I haven't seen it yet.

I think the middle ground is worthy of a little bit of discussion.

**Comment by kovidgoyal** [comment]
Note that RS is not a normal compiled language. It is a *trans*piled language. The slowdown from replacing == by either inlined code or a simple function call is very real. The reason for this is that the javascript VM already does a level of runtime type analysis and JITing based on that analysis. And crazy as it may seem, actual Javascript function length can determine whether JITing gets turned on for a function or not. So blowing up every == operator into dozens of lines of JS code is going to have an outsize effect. Not to mention that JS is not an assembly language. I know of no way to inline == comparisons in all contexts. You pretty much have to do a function call. 
Adding another layer of type analysis on top of that already done in javascript itself makes a much larger difference than you would intuitively expect. I know, because I actually spent the time to do this, not for == but for if (to make boolean evaluation of lists/dicts the same as in python). It caused a 700% slowdown in my tests on real world code.

I suggest you do some benchmarking for yourself, you are not going to be convinced otherwise. And if you discover a technique by which you can replace == with something that has python like semantics for only a few % slowdown, I will be delighted to implement it.

And while there are undoubtedly people that would love to use unmodified python in the browser, there are already lots of projects that do this. @atsepkov pointed you at several of them. The problem space that RS positions itself to solve is different, and trying to solve total python compatibility would move it out of that space. Language design, like most everything, is all about tradeoffs. If you believe in a different position on the efficiency/compatibility tradeoff, I suggest using one of those other languages.

**Comment by artyprog** [comment]
@JohnDenker 
Hell John, you can try the RapydScript version of GlowScript here :
http://salvatore.pythonanywhere.com/RapydGlow 
You can also try
http://salvatore.pythonanywhere.com/RapydBox
Regards

**Comment by atsepkov** [comment]
@JohnDenker first I apologize if I sound brash, that is not the intent. It's just that we already went through these philosophical discussions in the past, multiple times, and have made the mistake of relaxing our standards, multiple times, and every time it bit us, every time such feature had to be rolled back.

As mentioned earlier, I understand where you're coming from. Adding consistency of Python on top of JavaScript has always been the intent of RapydScript. But realize that the word **Rapyd** (rapid) is in the name for a reason, and performance is the reason GlowScript has adopted RapydScript as its language of choice - it's easy to add more complex blocks (operator overloading) on top of simpler ones, but it's harder to simplify a platform that already has bloat in it. As for GlowScript, where does it do operator overloading? Here is their RapydScript example (http://www.glowscript.org/#/user/GlowScriptDemos/folder/Examples/program/RotatingCubes-RapydScript/edit), the operations are performed on each coordinate individually.

Why don't you just do a global search for `==` and replace each occurrence with `eq`? That way you'll at least glance at each occurrence yourself rather than trusting an automated tool to handle safety? You're still expected to massage your code when porting it, don't be lazy.

Also the features you suggest for hinting the compiler to do the optimizations are possible but realize the following:
- What makes Python suck less than C++ is exactly that you don't need to litter your code with this
- This is not a simple feature, if you would like to tackle adding such optimizer yourself (https://github.com/atsepkov/RapydScript/issues/89) and then do a pull request, your help would be greatly appreciated

Finally, I once again want to reemphasize what I already mentioned in my first post here. I'm working on something that may get you closer to what you want, but you need to be patient until I finish merging other stability changes in from -ng fork (give me a week or two). And unlike typical feature requests, the long discussion here is eating into that time. For that reason, I'll refrain from responding to anything else in this thread until I'm done with tasks at hand - otherwise I'm doing a disfavor to both you and other members.

**Comment by atsepkov** [comment]
This thread is now irrelevant due to performant deep equality operator I added discussed here: http://blog.pyjeon.com/2015/12/07/equality-operator/
